### PR TITLE
Remove Radix cache path and consolidate runtime cache handling

### PR DESCRIPTION
**Pull request by wangbo981016**
## Summary

- Remove the legacy Radix cache path and related configuration, execution, and model plumbing.
- Consolidate cache management around the Flat/page-based runtime implementations.
- Update storage, transfer, PD, attention, model, container, CI, documentation, and test integrations accordingly.

## Testing

- Not run (not requested).

### CODE REVIEW

File: python/tokenspeed/runtime/entrypoints/engine.py [hunk 1]
```diff
@@ -606,4 +606,8 @@ def _launch_subprocesses(
     # Assume all schedulers have the same scheduler_info
     scheduler_info = scheduler_infos[0]
     tokenizer_manager.max_req_input_len = scheduler_info["max_req_input_len"]
+    tokenizer_manager.max_single_request_tokens = scheduler_info[
+        "max_single_request_tokens"
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/entrypoints/engine.py:637]: Good catch. Fixed in e7c3ddbd: the encode ready payload now includes max_single_request_tokens, matching the launcher contract before engine.py consumes the message.

File: python/tokenspeed/runtime/engine/event_loop.py [hunk 2]
```diff
@@ -290,91 +268,28 @@ def __init__(
...
-                    "the flat host tier (kvstore L2) does not support FlatKV "
-                    "contract pools (Kimi-K3) yet; pass --disable-kvstore."
+                    "the host tier does not support contract pools yet; pass "
+                    "--disable-kvstore"
                 )
-            self.memory_executor = FlatMemoryExecutor(
+            self.memory_executor = MemoryExecutor(
                 device_pool=token_to_kv_pool,
                 host_ratio=server_args.kvstore_ratio,
                 host_size_gb=server_args.kvstore_size,
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/engine/event_loop.py:306]: This is a valid issue: target-only host mirroring is insufficient for speculative decoding. L2/KVStore support is intentionally deferred in this PR; all CI serve configurations now pass --disable-kvstore. We will implement unified target+draft host transport as a separate follow-up, so I am leaving this thread unresolved.

File: python/tokenspeed/runtime/execution/block_table.py [hunk 3]
```diff
@@ -18,10 +18,30 @@
...
+            raise ValueError(
+                f"history_group_id {group_id!r} is missing from block tables "
+                f"{sorted(tables)}"
+            )
+        return tables[group_id]
+    if len(tables) != 1:
+        raise ValueError(
+            "history_group_id is required when the scheduler exports "
+            f"{len(tables)} block tables"
+        )
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/execution/block_table.py:39]: This is valid for the legacy multi-group DeepSeek V4 transfer path. DeepSeek V4 and L2 adaptation are explicitly out of scope for this PR and will be implemented against the consolidated cache contract separately, so I am leaving this thread unresolved.

File: tokenspeed-scheduler/csrc/scheduler/operations/forward.cpp [hunk 4]
```diff
@@ -161,1267 +87,444 @@ bool canConsumeAvailable(const KvCacheCoordinator& coordinator, std::span<const
...
+                request_to_retract = request;
             }
         }
     }
 
-    op.begin = begin;
-    op.size = current_size - begin;
-    previous.prefix_pages = current_prefix;
-    previous.local_pages.assign(op.occupied_pages.begin() + current_prefix, op.occupied_pages.end());
+    FatalCheck(request_to_retract != nullptr, "LCM admission failed without a retractable request");
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/operations/forward.cpp:429]: Verified. The single-request capacity bound does not protect two incomplete prefills when a promoted boundary leaves enough token budget to start the second request. Fixed in c5cbc979: an incomplete prefill remains head-of-line, so no second Submitted request can acquire pages until it completes. Added a regression that reproduces the full-history/state promotion-boundary case; all 346 scheduler C++ tests pass.

File: python/tokenspeed/runtime/cache/host_mirror.py [hunk 5]
```diff
@@ -47,10 +47,10 @@ def _state_slabs(device_kv_pool) -> list[tuple[torch.Tensor, torch.Tensor]]:
...
 
-def flat_bytes_per_host_page(device_kv_pool) -> int:
+def bytes_per_host_page(device_kv_pool) -> int:
     """Bytes one host page occupies across all mirrors, computed from the
     device pool alone (no mirror allocation) -- the sizing side of
-    ``FlatHostMirror.bytes_per_host_page`` for host-budget arithmetic.
+    ``HostMirror.bytes_per_host_page`` for host-budget arithmetic.
     """
     tensors = _identity_dedup(device_kv_pool.k_buffer) + _identity_dedup(
         device_kv_pool.v_buffer
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/cache/host_mirror.py:56]: Valid for the current HostMirror implementation: standard MLA pools expose kv_buffer, not the MHA k_buffer/v_buffer pair. L2/KVStore support is explicitly out of scope for this PR; all affected CI serve commands now pass --disable-kvstore. We will add MLA-aware transport (or a fail-fast capability rejection) with the unified L2 implementation, so I am leaving this thread open.

File: python/tokenspeed/runtime/engine/event_loop.py [hunk 6]
```diff
@@ -314,93 +268,28 @@ def __init__(
...
-                # raw-slab pool does not expose. Reject this unsupported
-                # combination at startup.
                 raise NotImplementedError(
-                    "the flat host tier (kvstore L2) does not support FlatKV "
-                    "contract pools (Kimi-K3) yet; pass --disable-kvstore."
+                    "the host tier does not support contract pools yet; pass "
+                    "--disable-kvstore"
                 )
-            self.memory_executor = FlatMemoryExecutor(
+            self.memory_executor = MemoryExecutor(
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/engine/event_loop.py:303]: Valid: a pool that explicitly opts out of hierarchical caching must not fall through to the generic mirror, and MSA has additional index state that the current mirror does not restore. L2/KVStore is intentionally deferred from this PR and the CI paths disable it with --disable-kvstore. This should be handled by the follow-up L2 capability/descriptor work, so I am leaving the thread open.

File: python/tokenspeed/runtime/cache/executor/memory_executor.py [hunk 7]
```diff
@@ -18,585 +18,384 @@
...
-                io_backend=config.io_backend,
-                layer_num=actual_pool.layer_num,
-                draft_device_pool=(
-                    actual_draft_pool if draft_device_pool is not None else None
-                ),
-                draft_host_pool=self.draft_host_pool,
-                draft_layer_num=draft_layer_num,
+        requested_bytes = host_page_count * host_page_bytes
+        available_bytes = psutil.virtual_memory().available - _HOST_MEM_HEADROOM_BYTES
+        if requested_bytes > available_bytes:
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/cache/executor/memory_executor.py:150]: Valid for L2: the host-pinned allocation budget must be divided across local workers and respect the container/cgroup limit. L2/KVStore is explicitly out of scope here and all affected CI paths now use --disable-kvstore; the multi-worker budget will be restored in the follow-up unified L2 implementation. Leaving this thread open.

File: python/tokenspeed/runtime/configs/paged_cache_spec.py [hunk 8]
```diff
@@ -191,117 +143,85 @@ def hybrid_slab_group_size(
...
         )
     if speculative_algorithm == "DFLASH":
-        raise RuntimeError(
-            "flat scheduler build (TOKENSPEED_FLAT_KVCACHE): DFLASH block "
-            "decode is unsupported on the flat path. Use a radix-built "
-            "tokenspeed_scheduler extension."
-        )
-    if len(paged_cache_groups) > 1 and not uses_flat and contract is None:
+        raise RuntimeError("DFLASH is not supported by the cache-group scheduler")
+    if len(paged_cache_groups) > 1 and not uses_cache_groups and contract is None:
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/configs/paged_cache_spec.py:190]: Valid for the DeepSeek V4 backend: uses_paged_cache_groups is its grouped-table capability and this validation currently only reads uses_cache_groups. DeepSeek V4 is explicitly out of scope for this PR, along with its L2 path, so I am not adding a partial compatibility exception here. We will unify the capability flags in the DSV4 follow-up; leaving this thread open.

File: tokenspeed-scheduler/csrc/scheduler/scheduler.cpp [hunk 9]
```diff
@@ -24,571 +24,322 @@
...
     : config_{std::move(config)},
-      device_allocator_{config_.block_size, config_.device_allocator.total_pages},
-      host_allocator_{config_.block_size, config_.host_allocator.total_pages},
-      mamba_allocator_{},
-      kv_prefix_cache_{&device_allocator_, &host_allocator_, config_.enable_l3_storage, config_.disable_prefix_cache},
-      req_pool_allocator_{config_.max_batch_size}
-#if TOKENSPEED_FLAT_KVCACHE
-      ,
+      req_pool_allocator_{config_.max_batch_size},
       block_pool_{config_.device_allocator.total_pages - 1},
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/scheduler.cpp:65]: Fixed in feaf5804. The fallback token count describes usable pages, while the scheduler total must also include reserved null page 0, so ordinary pools now report `pages + 1` device pages and keep `pages` usable pages. Added a boundary test for the 4-usable-page case.

File: tokenspeed-scheduler/csrc/scheduler/scheduler.cpp [hunk 10]
```diff
@@ -24,571 +24,322 @@
...
     }
-    const auto& hashes = result.host.last_node->PageHashes();
-    std::string prior = hashes.empty() ? std::string{} : hashes.back();
+    return static_cast<std::size_t>(std::distance(cache_group_ids_.begin(), it));
+}
 
-    return ComputePagedHashes(
-        std::vector<std::span<const std::int32_t>>(token_pages.begin() + host_matched, token_pages.end()), prior);
+std::vector<KvCacheEvent> Scheduler::DrainKvEvents() {
+    return std::exchange(kv_events_, {});
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/scheduler.cpp:216]: Fixed in 7dc86841. The coordinator now reports real device-cache insertions/removals, and the scheduler folds per-group mutations into one externally visible BlockStored/BlockRemoved event. Added coverage for stored, removed, and multi-group publication.

File: tokenspeed-scheduler/csrc/fsm/forward_events.cpp [hunk 11]
```diff
@@ -18,681 +18,105 @@
...
+    return Decoding{token_container,           page_size,
+                    std::move(req_pool_index), decode_input_tokens_,
+                    std::move(block_tables),   std::move(cache_progress_)};
+}
+
+template <typename State>
+Finished FinishEvent::finish(State&& state) {
+    _assert(coordinator_ != nullptr, "FinishEvent requires a cache coordinator");
+    auto block_tables = std::move(state).TakeBlockTables();
+    FreeRequest(*coordinator_, block_tables);
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/fsm/forward_events.cpp:110]: Fixed in eda6bbf0. Finish now hashes and publishes every stable page completed by the last forward before releasing the request tables. The coordinator reuses the existing per-group completion policy, so Full KV is registered densely while State/SWA still honor endpoint/checkpoint rules. Added PrefixHitSuite.FinishPublishesPagesFromLastForward; all 352 scheduler C++ tests pass.

File: python/tokenspeed/runtime/configs/paged_cache_spec.py [hunk 12]
```diff
@@ -191,117 +143,85 @@ def hybrid_slab_group_size(
...
+            "with speculative decoding"
         )
     if speculative_algorithm == "DFLASH":
-        raise RuntimeError(
-            "flat scheduler build (TOKENSPEED_FLAT_KVCACHE): DFLASH block "
-            "decode is unsupported on the flat path. Use a radix-built "
-            "tokenspeed_scheduler extension."
-        )
-    if len(paged_cache_groups) > 1 and not uses_flat and contract is None:
+        raise RuntimeError("DFLASH is not supported by the cache-group scheduler")
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/configs/paged_cache_spec.py]: Fixed in f34c4125. DFLASH now passes the normal speculative capability checks, and its block-decode drafter stays on its independent single-table req_to_page path instead of capturing target cache-group buffers. Added focused DFLASH-vs-EAGLE graph-contract coverage.

File: python/tokenspeed/runtime/engine/event_loop.py [hunk 13]
```diff
@@ -314,93 +268,28 @@ def __init__(
...
-        )
-        paged_cache_host_group_pages = {}
-        if scheduler_ext_flat_kvcache() and server_args.enable_kvstore:
+        if server_args.enable_kvstore:
             if server_args.kvstore_storage_backend is not None:
                 raise NotImplementedError(
-                    "flat scheduler build (TOKENSPEED_FLAT_KVCACHE) has no L3 "
-                    "storage tier yet; unset --kvstore-storage-backend."
+                    "the cache-group scheduler has no L3 storage tier; unset "
+                    "--kvstore-storage-backend"
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/engine/event_loop.py:296]: Valid for the removed L2/L3 transport stack. L2/KVStore/L3 restoration is explicitly deferred from this scheduler-consolidation PR; affected CI runs with --disable-kvstore. We will restore Mooncake through the unified descriptor/transport follow-up, so this remains open rather than adding a partial path here.

**Comment by wangbo981016** [comment]
Addressed the chunk-alignment P1 in f34c4125: state-page alignment is now resolved before ModelExecutorConfig/InputBuffers construction. Downward alignment updates server_args before sizing; a configured chunk smaller than one recurrent-state page is rejected instead of silently increasing a frozen buffer limit. Added/updated boundary coverage.

File: python/tokenspeed/runtime/pd/prefill_executor.py [hunk 14]
```diff
@@ -380,8 +357,8 @@ def abort(self, request_id: str, bootstrap_info: BootstrapInfo) -> None:
         (embedding receive timed out). Signal the dual-dispatched decode so its KV
         receiver fails instead of waiting forever. No sender was registered, so
         there is nothing to tear down on this side."""
-        if self.is_flatkv:
-            raise NotImplementedError("FlatKV PD does not support EPD admission")
+        if self.uses_cache_contract:
+            raise NotImplementedError("Paged cache PD does not support EPD admission")
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/pd/prefill_executor.py]: Fixed in 28dbf90e. Cache-contract EPD aborts now call `kv_manager.abort_room(...)`, so the decode endpoint receives the transfer failure instead of waiting for its longer timeout. Added `test_cache_contract_epd_abort_notifies_decode`; the full `test_cache_pd_executors.py` suite passes (11/11).

File: tokenspeed-scheduler/csrc/scheduler/execution_plan.h [hunk 15]
```diff
@@ -23,73 +23,29 @@
...
-    ExecutionPlan& With(std::vector<OperationType> ops) {
-        for (auto& op : ops) {
-            operations_.emplace_back(std::move(op));
-        }
-        return *this;
-    }
-
-    ExecutionPlan& WithSchedulerAborts(std::vector<SchedulerAbort> aborts) {
-        scheduler_aborts_ = std::move(aborts);
+        operations_.emplace_back(operation);
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/execution_plan.h]: Fixed in 7bf88238. `ExecutionPlan::With` now moves its by-value operation into the variant, avoiding a deep copy of `ForwardBatch` and its per-group page tables. Rebuilt the scheduler and ran all 352 C++ tests; full pre-commit also passes.

File: tokenspeed-scheduler/csrc/scheduler/scheduler.h [hunk 16]
```diff
@@ -28,159 +28,99 @@
...
 
     ExecutionPlan NextExecutionPlan();
-
     void Advance(const ExecutionEvent& event);
     std::vector<KvCacheEvent> DrainKvEvents();
 
     std::size_t WaitingSize() const;
     std::size_t DecodingSize() const;
-    std::size_t RetractedSize() const;
+    std::size_t RetractedSize() const { return static_cast<std::size_t>(retract_count_); }
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/scheduler.h]: Fixed in 5ba2705d. In the canonical scheduler, RetractEvent immediately transitions Decoding/PrefillDone back to Submitted, so WaitingSize already counts every live retracted request. I removed the cumulative retract_count_, its Python binding, and the redundant drain predicate instead of inventing a second lifecycle count. Scheduler C++ tests pass 352/352 and pre-commit passes.

File: python/tokenspeed/runtime/execution/model_executor.py [hunk 17]
```diff
@@ -1494,9 +1300,10 @@ def update_block_table(self, forward_op) -> ModelExecutionResult:
                 forward_op=forward_op,
                 device=self.device,
                 req_to_page=self.req_to_page,
+                history_group_id=self._full_history_group_id,
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply python/tokenspeed/runtime/execution/model_executor.py:1365]: Fixed in bce81e64. ModelExecutor now skips the legacy req_to_page mirror when a runtime cache contract has no full-history group, because the group-aware backend consumes CacheBatchMetadata directly and there is no unambiguous compatibility table. Added a regression test for this case.

File: tokenspeed-scheduler/csrc/scheduler/outside_event_handler.cpp [hunk 18]
```diff
@@ -18,314 +18,164 @@
...
     pending_forward_results_.erase(event.request_id);
-#if TOKENSPEED_FLAT_KVCACHE
-    flat_pd_transfer_pins_.erase(event.request_id);
-#endif
-    Request* req = find_request(event.request_id);
-    if (req == nullptr) {
-        return;
+    pd_transfer_pins_.erase(event.request_id);
+    if (Request* request = findRequest(event.request_id)) {
+        request->Apply(fsm::AbortEvent{&coordinator_});
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/outside_event_handler.cpp:153]: This is intentionally safe in the current overlap pipeline. The scheduler may release CPU ownership after committing the previous result, but the next iteration first orders the default stream after ModelExecutor.execution_stream; page-table updates and page zeroing therefore cannot execute until the already-launched forward has finished. execute_forward_op then orders execution_stream back after those default-stream writes. The delayed result is deliberately ignored after its Python RequestState is removed. An independently scheduled L2/DMA transport would require its own fence, but that path is disabled and deferred in this PR. Deferring terminalization here would add state without changing GPU ordering, so no code change is needed.

File: tokenspeed-scheduler/csrc/scheduler/outside_event_handler.cpp [hunk 19]
```diff
@@ -18,314 +18,164 @@
...
-    request->Apply(fsm::FinishEvent{&kv_prefix_cache_, &host_allocator_, std::move(page_hashes),
-                                    config_.disable_l2_cache, hybrid_prefix_cache_ ? &*hybrid_prefix_cache_ : nullptr
-#if TOKENSPEED_FLAT_KVCACHE
-                                    ,
-                                    &coordinator_
-#endif
-    });
-    tryFinalizeDeferredAbort(event.request_id);
+    pd_transfer_pins_.erase(event.request_id);
+    request->Apply(fsm::FinishEvent{&coordinator_});
```
[the remark this answers is not in the document]
Comment (wangbo981016) [reply tokenspeed-scheduler/csrc/scheduler/outside_event_handler.cpp:75]: The P-side prefix is already published before SucceededEvent can exist. A prefill worker receives SucceededEvent only after the zero-token transfer operation has been scheduled. Building that operation goes through Scheduler::scheduleDecode(), which advances completed page hashes and passes the endpoint boundary into Admit(); KvCacheCoordinator::Admit() calls cacheCompletedBlocksForGroup before the transfer is emitted. SucceededEvent is therefore only the post-transfer lifetime release. Calling publishCompletedPages again would be redundant, and a sub-page prompt has no stable page to publish.
